temporalio/worker/workflow_sandbox/runner.py

```diff
@@ -0,0 +1,154 @@
+        return _Instance(det, self._runner_class, self._restrictions)
+
+
+# Implements in_sandbox._ExternEnvironment. Some of these calls are called from
+# within the sandbox.
+class _Instance(WorkflowInstance):
+    def __init__(
+        self,
+        instance_details: WorkflowInstanceDetails,
+        runner_class: Type[WorkflowRunner],
+        restrictions: SandboxRestrictions,
+    ) -> None:
+        self.instance_details = instance_details
+        self.runner_class = runner_class
+        self.importer = Importer(restrictions, RestrictionContext())
+
+        # Create the instance
+        self.globals_and_locals = {
+            "__file__": "workflow_sandbox.py",
+        }
+        self._create_instance()
+
+    def _create_instance(self) -> None:
+        try:
+            # Import user code
+            self._run_code(
+                "with __temporal_importer.applied():\n"
+                # Import the workflow code
+                f"  from {self.instance_details.defn.cls.__module__} import {self.instance_details.defn.cls.__name__} as __temporal_workflow_class\n"
+                f"  from {self.runner_class.__module__} import {self.runner_class.__name__} as __temporal_runner_class\n",
+                __temporal_importer=self.importer,
+            )
+
+            # Set context as in runtime
+            self.importer.restriction_context.is_runtime = True
+
+            # Create the sandbox instance
+            self._run_code(
+                "with __temporal_importer.applied():\n"
+                "  from temporalio.worker.workflow_sandbox.in_sandbox import InSandbox\n"
+                "  __temporal_in_sandbox = InSandbox(__temporal_instance_details, __temporal_runner_class, __temporal_workflow_class)\n",
+                __temporal_importer=self.importer,
+                __temporal_instance_details=self.instance_details,
+            )
+        finally:
+            self.importer.restriction_context.is_runtime = False
+
+    def activate(
+        self, act: temporalio.bridge.proto.workflow_activation.WorkflowActivation
+    ) -> temporalio.bridge.proto.workflow_completion.WorkflowActivationCompletion:
+        self.importer.restriction_context.is_runtime = True
+        try:
+            self._run_code(
+                "with __temporal_importer.applied():\n"
+                "  __temporal_completion = __temporal_in_sandbox.activate(__temporal_activation)\n",
+                __temporal_importer=self.importer,
+                __temporal_activation=act,
+            )
+            return self.globals_and_locals.pop("__temporal_completion")  # type: ignore
+        finally:
+            self.importer.restriction_context.is_runtime = False
+
+    def _run_code(self, code: str, **extra_globals: Any) -> None:
+        for k, v in extra_globals.items():
+            self.globals_and_locals[k] = v
+        try:
+            exec(code, self.globals_and_locals, self.globals_and_locals)
```

Comment:
I know we didn't test for performance, but do you think compiling and caching the workflow code could be beneficial?

Reply:
Probably not much. This runs only a few lines of code. Two of those are static yes (another is dynamic), but the actual user's workflow code is already cached by Python in the loader and import just re-executes it. The `compile()` step here would only compile those few lines. But yes, upon benchmarking, this may be worth it. I can make that code static too. I may come back to this after profiling.